### PR TITLE
[WIP] Avoid object tree traversal

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
Here is what I have so far.
I had to add a dep to one of my nugets that handles some implicit reflection stuff. (cause reflectively invoking properties / methods is really annoying)

### CODE REVIEW

File: Saule/Saule.csproj [hunk 1]
```diff
@@ -41,6 +41,9 @@
     <CodeAnalysisIgnoreGeneratedCode>true</CodeAnalysisIgnoreGeneratedCode>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="csharp-extensions">
+      <HintPath>D:\ClientDevelopment\AllisonTransmission\Integration.API\packages\csharp-extensions.1.2.5\lib\net46\csharp-extensions.dll</HintPath>
```
Comment (NullVoxPopuli) [inline Saule/Saule.csproj]: ha, this shouldn't be in here.

File: Saule/Serialization/ResourceSerializer.cs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Linq;
+using csharp_extensions.Extensions;
```
Comment (NullVoxPopuli) [inline Saule/Serialization/ResourceSerializer.cs]: this is the dep

File: Saule/Serialization/ResourceSerializer.cs [hunk 3]
```diff
@@ -66,71 +69,80 @@ public JObject Serialize(JsonSerializer serializer)
...
+        private static string EnsureHasId(object properties, ApiResource resource)
         {
             var id = GetId(properties, resource);
             if (id == null)
             {
                 throw new JsonApiException(ErrorType.Server, "Resources must have an id");
             }
 
-            return id;
+            return id.ToString();
```
Comment (NullVoxPopuli) [inline Saule/Serialization/ResourceSerializer.cs]: It's better to do `.ToString()` here than to cast to string, cause of weird primary key types (like guids --  /cc @wozbo)

**Comment by joukevandermaas** [comment]
@NullVoxPopuli it looks like you only added the reference locally, because it is not found on the build server. Do you think you'll be able to wrap up this PR? It looks pretty good so far.

**Comment by NullVoxPopuli** [comment]
@joukevandermaas I don't think I'll be able to -- I'm no longer working with C# (switched jobs).
If you want, I can give you write permissions to my fork, though

**Pull request closed without merging**